src/nerve_tools/envoy.py

```diff
@@ -52,10 +54,26 @@ def get_envoy_ingress_listeners(admin_port: int) -> Mapping[int, int]:
 def get_envoy_service_info(
     service_name: str,
     service_info: ServiceInfo,
-    envoy_ingress_listeners: Mapping[int, int],
+    envoy_ingress_listeners: Mapping[Tuple[str, int], int],
 ) -> Optional[ServiceInfo]:
     envoy_service_info: Optional[ServiceInfo] = None
-    local_host_port = service_info['port']
+
+    # Previously, mesos managed services running on this host had unique ports so
+    # mesos port -> envoy ingress port mapping was pretty straight forward. With
+    # Kubernetes services, the port is not guaranteed to be unique to this host
+    # because of the pod abstraction which introduces a pod ip address. Thus, to
+    # maintain a valid mapping, a composite key (service_name, service_port) is used
+    # to map to the service's envoy ingress port.
+    #
+    # This is not fullproof! multiple instances of the same service may be scheudled
+    # to run on this host. The (service_name, service_ip, service_port) is the only
+    # key that would maintain uniqueness, but service_ip (the pod_id in the case of k8s)
+    # is not part of the information returned from the envoy admin port /listener
+    # response.
+
+    # TODO(spatel) Figure out a tighter solution here
+    key = (service_name, service_info['port'])
```

Comment:
@kaisen Wanted to make sure I'm not missing something here before I go spend some more time on this.

Reply:
this sounds right to me